src/coordinator/ts/m3db/block/iterator.go

```diff
@@ -0,0 +1,214 @@
+// //
+// // THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// // IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// // FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// // AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// // LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// // THE SOFTWARE.
+
+package block
+
+import (
+	"errors"
+	"math"
+	"time"
+
+	"github.com/m3db/m3db/src/coordinator/block"
+)
+
+// StepIter creates a new step iterator for a given MultiSeriesBlock
+func (m MultiSeriesBlock) StepIter() (block.StepIter, error) {
+	return &multiSeriesBlockStepIter{
+		seriesIters: newConsolidatedSeriesBlockIters(m.Blocks),
+		index:       -1,
+		meta:        m.Metadata,
+		blocks:      m.Blocks,
+	}, nil
+}
+
+// SeriesIter creates a new series iterator for a given MultiSeriesBlock
+func (m MultiSeriesBlock) SeriesIter() (block.SeriesIter, error) {
+	// todo(braskin): implement SeriesIter()
+	return nil, errors.New("SeriesIter not implemented")
+}
+
+// Close frees up resources
+func (m MultiSeriesBlock) Close() error {
+	// todo(braskin): Actually free up resources
+	return nil
+}
+
+func newConsolidatedSeriesBlockIters(blocks ConsolidatedSeriesBlocks) []block.ValueIterator {
+	consolidatedSeriesBlockIters := make([]block.ValueIterator, len(blocks))
+	if len(blocks) == 0 {
+		return consolidatedSeriesBlockIters
+	}
+
+	for i, seriesBlock := range blocks {
+		consolidatedNSBlockIters := make([]block.ValueIterator, len(blocks[0].ConsolidatedNSBlocks))
+		for j, nsBlock := range seriesBlock.ConsolidatedNSBlocks {
+			nsBlockIter := newConsolidatedNSBlockIter(nsBlock)
+			consolidatedNSBlockIters[j] = nsBlockIter
+		}
+		consolidatedSeriesBlockIters[i] = &consolidatedSeriesBlockIter{
+			consolidatedNSBlockIters: consolidatedNSBlockIters,
+		}
+	}
+
+	return consolidatedSeriesBlockIters
+}
+
+func newConsolidatedNSBlockIter(nsBlock ConsolidatedNSBlock) *consolidatedNSBlockIter {
+	return &consolidatedNSBlockIter{
+		consolidatedNSBlockSeriesIters: nsBlock.SeriesIterators.Iters(),
+		bounds:    nsBlock.Bounds,
+		indexTime: nsBlock.Bounds.Start.Add(-1 * nsBlock.Bounds.StepSize),
+	}
+}
+
+// Meta returns the metadata for the step iter
+func (m *multiSeriesBlockStepIter) Meta() block.Metadata {
+	return m.meta
+}
+
+// SeriesMeta returns metadata for the individual timeseries
+func (m *multiSeriesBlockStepIter) SeriesMeta() []block.SeriesMeta {
+	metas := make([]block.SeriesMeta, len(m.blocks))
+	for i, s := range m.blocks {
+		// todo(braskin): add name
+		metas[i].Tags = s.Metadata.Tags
+	}
+	return metas
+}
+
+// StepCount returns the total steps/columns
+func (m *multiSeriesBlockStepIter) StepCount() int {
+	if len(m.blocks) == 0 {
+		return 0
+	}
+	return m.blocks[0].Metadata.Bounds.Steps()
```

Comment:
We should probably have  methods on `block` that will return things like Metadata and Steps so we don't have to do this everywhere... though this is a bigger discussion than for here )